### PR TITLE
[SPARK-47318][CORE] Adds HKDF round to AuthEngine key derivation to follow standard KEX practices

### PR DESCRIPTION
**Pull request by sweisdb**
### What changes were proposed in this pull request?

This change adds an additional pass through a key derivation function (KDF) to the key exchange protocol in `AuthEngine`. Currently, it uses the shared secret from a bespoke key negotiation protocol directly. This is an encoded X coordinate on the X25519 curve. It is atypical and not recommended to use that coordinate directly as a key, but rather to pass it to an KDF.

Note, Spark now supports TLS for RPC calls. It is preferable to use that rather than the bespoke AES RPC encryption implemented by `AuthEngine` and `TransportCipher`.

### Why are the changes needed?

This follows best practices of key negotiation protocols. The encoded X coordinate is not guaranteed to be uniformly distributed over the 32-byte key space. Rather, we pass it through a HKDF function to map it uniformly to a 16-byte key space.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Exiting tests under:
`build/sbt "network-common/test:testOnly"`

Specifically:
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthEngineSuite"`
`build/sbt "network-common/test:testOnly org.apache.spark.network.crypto.AuthIntegrationSuite"`

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @vanzin, @srowen 

Can you also include the documentation change in the PR to call out the incompatibility introduced due to this ? Thanks

**Review by dongjoon-hyun — COMMENTED** [review]
Could you add a reference link for this statement in the PR description, @sweisdb ?
> This follows best practices of key negotiation protocols.

**Comment by sweisdb** [comment]
@dongjoon-hyun Will do. I would clarify that the best practice here would be to use TLS and not use a bespoke key exchange protocol. 

But if we are going to use a bespoke key exchange protocol, we need to run the output of the key exchange through a KDF.

**Comment by sweisdb** [comment]
I've updated the overall Security doc to indicate that TLS is preferred to using the legacy RPC encryption. I also updated the AuthEngine README to talk about incompatibility.

I retroactively made up v1.0 and v1.1 of the key exchange protocol to talk about this. It would probably be more understandable to talk about Spark versions. Should we talk about 4.0 in the context of this change and amend the docs if and when it's backported?

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:12]: I am not sure if we should announce deprecation given TLS is going to come out only in 4.0 ... we can recommend users to move to TLS though.

Would like feedback from others on this though ... @dongjoon-hyun, @srowen, @JoshRosen

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:123]: Essentially, this is backwardly incompatible with 3.x  on the wire - making adoption of 4.x in secure env all-or-nothing
Given this, I am leaning towards flag guarding the change.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:223]: Let us flag guard this change - given we want to backport to 3.x patch releases as well as giving users migration opportunities. This flag can be flipped to default in a subsequent minor version.

**Review by mridulm — COMMENTED** [review]
For context to reviewers, the plan is to backport this to 3.x as well given security implications - that is, we will will be making an incompatible wire protocol change.

Please review from that context as well - thanks !

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:123]: +1 for @mridulm 's recommendation.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:126]: I assume that Apache Spark community follows `Semantic Versioning` by default. If this is incompatible, shall we use `v2.0` instead of `v1.1`?

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:1]: `v2.0`? (https://github.com/apache/spark/pull/45425/files/4204dd1f03cb86f4f6e082d2c4b8bd5f402ad12f..7ace803c40ab50c03e4af1c3733499668fb340d5#r1525394252)

File: docs/security.md
Comment (dongjoon-hyun) [inline docs/security.md:173]: Since this layer is completely irrelevant to UI, we don't need to say like this. Shall we omit this paragraph?
> Unlike the other SSL settings for the UI

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:8]: To @sweisdb , may I ask when you think we can delete this? If you think `Spark 5`, please remove this claim. It's too far. In Apache Spark community, deprecation doesn't mean the removal.
> and removed in the long-term.

FYI, here is the guideline.
- https://spark.apache.org/versioning-policy.html
https://spark.apache.org/versioning-policy.html

**Review by dongjoon-hyun — COMMENTED** [review]
BTW, from my side, we don't have a mixed Spark-version RPC use case. So, it's okay to introduce this if we have the fallback configuration to disable this completely (as recommended by Mridul).

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:12]: Given that SASL support still exists for backward compatibility, I am skeptical that this will actually be removed. I'll delete that comment.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:1]: Will fix.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:8]: After seeing that we're still supporting legacy SASL, I agree that this will likely not be removed any time in the forseeable future.

**Comment by dongjoon-hyun** [comment]
Got it, @mridulm .
> For context to reviewers, the plan is to backport this to 3.x as well given security implications - that is, we will will be making an incompatible wire protocol change.

If this is for Apache Spark 3.5.2 and 3.4.3, cc @cloud-fan , @HyukjinKwon , too

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/README.md]: Fixing this.

**Comment by sweisdb** [comment]
I had a bad push that pulled in a lot of other unrelated commits. Will fix.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:228]: I added a flag to let someone override this and revert back to the original protocol.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:219]: Make the default 1, to preserve the existing behavior, and allowing backport to patch versions.
We can document to set this explicitly to to `2` to mitigate this issue.

In a later minor version, we can flip this to `2` by default.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:239]: Let us minimize optional changes in this PR (this by itself is fine - but in general, we should reduce the scope of change), given we want to backport to older versions and want to minimize risk.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (mridulm) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:109]: nit:

```suggestion
The original version of this protocol, retroactively called v1.0, did not apply an HKDF to `sharedSecret` (to generate the `derivedKey`) and was
```

File: docs/security.md
Comment (mridulm) [inline docs/security.md:160]:
```suggestion
RPC authentication to also be enabled. This method uses a bespoke protocol with an older v1 and an updated v2 version. <add details about v1 vs v2>
```

Also, update the config section below on how to configure `spark.network.crypto.authEngineVersion`, and a ref to that here: along with incompatibility between `1` and `2` (and all versions prior being `1` retroactively - without a way to configure it before `3.5.2` and `3.4.3`)

**Review by mridulm — COMMENTED** [review]
+CC @dongjoon-hyun as well.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java:239]: +1 for @mridulm 's comment. This should be reverted.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java
Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:219]: +1 for the above comment, too.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java]: Reverted. Good call.

File: docs/security.md
Comment (sweisdb) [reply docs/security.md:160]: Updated in the section about the Legacy protocol.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:217]: `Value values`?

Comment (dongjoon-hyun) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java:217]: Also, shall we mentioned that the default is 1 intentionally?

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:111]: `The current version, v2.0,` could mislead the users. Let's say `The latest version` instead.

File: common/network-common/src/main/java/org/apache/spark/network/crypto/README.md
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/crypto/README.md:115]: Shall we add a sentence to describe Apache Spark sticks to `v1.0` by default due to the above reason?

File: common/network-common/src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java
Comment (dongjoon-hyun) [inline common/network-common/src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:64]: I'm wondering if this means that we don't test `v1` from now.

We had better keep the test coverage for v1 in addition to v2.

File: docs/security.md
Comment (dongjoon-hyun) [inline docs/security.md:188]: Thank you for the above description.

File: docs/security.md
Comment (dongjoon-hyun) [inline docs/security.md:208]: `Added in 4.0.0` looks like a typo. This will break this HTML table. In other words, we need to have two lines of `<td></td>`. One for description and the other for version. Please remove ` Added in 4.0.0` from here and create a new line, `<td>4.0.0</td>`, in order to follow the other configuration's style.

File: common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java
Comment (sweisdb) [reply common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java]: Good catch. Done.

File: common/network-common/src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java
Comment (sweisdb) [reply common/network-common/src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:64]: We test both v1 and v2. See below where I specifically give it a config with "1".

File: docs/security.md
Comment (sweisdb) [reply docs/security.md]: Good catch. Fixed.

File: docs/security.md
Comment (mridulm) [inline docs/security.md:160]: Looking at the earlier text, we are dropping some verbiage from it - any particular reason for it ?
Specifically `"However, the AES based encryption is simpler to configure and may be preferred if the only requirement is that data be encrypted in transit."` for ex.

We have made our recommendation clear, and rest should be left to users to determine - and make the right tradeoff.

**Review by mridulm — COMMENTED** [review]
Just one comment, rest looks good to me.

+CC @JoshRosen as well to have an extra pair of eyes to review.

**Comment by mridulm** [comment]
Can you please fix the linter related build errors @sweisdb ? Thx

**Comment by mridulm** [comment]
Btw, we would not be able to directly merge this PR into older versions - given the reference to SSL in the docs.
So unfortunately, we will need a follow up PR for 3.5.

File: docs/security.md
Comment (sweisdb) [reply docs/security.md:160]: I will take this out. I think it is rationalizing why there is an ad hoc protocol here instead in the first place. I agree that we should make it clear that TLS is strongly preferred and make it easy to migrate.

Comment (sweisdb) [reply docs/security.md:160]: I put the original text back in.

**Review by mridulm — COMMENTED** [review]
Looks good to me, thanks for working on this @sweisdb !
+CC @dongjoon-hyun

**Review by mridulm — CHANGES_REQUESTED** [review]
The java linter failures are still there, can you please fix those ?
See: https://github.com/sweisdb/spark/actions/runs/8603440358/job/23575468152

**Review by dongjoon-hyun — COMMENTED** [review]
Ya, please use `dev/lint-java` script and fix the following lines, @sweisdb .
```
$ dev/lint-java
Using `mvn` from path: /opt/homebrew/bin/mvn
Using SPARK_LOCAL_IP=localhost
Checkstyle checks failed at following occurrences:
[ERROR] src/main/java/org/apache/spark/network/crypto/AuthEngine.java:[68] (sizes) LineLength: Line is longer than 100 characters (found 113).
[ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[218] (sizes) LineLength: Line is longer than 100 characters (found 116).
[ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[220] (sizes) LineLength: Line is longer than 100 characters (found 101).
[ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[54] (sizes) LineLength: Line is longer than 100 characters (found 110).
[ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[55] (sizes) LineLength: Line is longer than 100 characters (found 116).
```

**Comment by sweisdb** [comment]
> Ya, please use `dev/lint-java` script and fix the following lines, @sweisdb .
> 
> ```
> $ dev/lint-java
> Using `mvn` from path: /opt/homebrew/bin/mvn
> Using SPARK_LOCAL_IP=localhost
> Checkstyle checks failed at following occurrences:
> [ERROR] src/main/java/org/apache/spark/network/crypto/AuthEngine.java:[68] (sizes) LineLength: Line is longer than 100 characters (found 113).
> [ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[218] (sizes) LineLength: Line is longer than 100 characters (found 116).
> [ERROR] src/main/java/org/apache/spark/network/util/TransportConf.java:[220] (sizes) LineLength: Line is longer than 100 characters (found 101).
> [ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[54] (sizes) LineLength: Line is longer than 100 characters (found 110).
> [ERROR] src/test/java/org/apache/spark/network/crypto/AuthEngineSuite.java:[55] (sizes) LineLength: Line is longer than 100 characters (found 116).
> ```

Addressed the lint issues.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs). Thank you!

**Comment by sweisdb** [comment]
Looks like the CI build is passing: https://github.com/sweisdb/spark/actions/runs/8623329437

**Comment by dongjoon-hyun** [comment]
Ya, it looks like that, @sweisdb .

**Comment by dongjoon-hyun** [comment]
Hi, @mridulm . Could you switch your request change?

**Comment by dongjoon-hyun** [comment]
BTW, @sweisdb , could you make two backporting PR to `branch-3.5` and `branch-3.4`?

I'm the release manager for Apache Spark 3.4.3 and targeting RC1 next Monday. I want to deliver this patch via Apache Spark 3.4.3.

**Comment by sweisdb** [comment]
@dongjoon-hyun:
v3.5.0 backport here: https://github.com/apache/spark/pull/46014
v3.4.0 backport here: https://github.com/apache/spark/pull/46015

**Comment by dongjoon-hyun** [comment]
Thank you so much!

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Merged to master.
Thanks for fixing this @sweisdb !
Thanks for reviewing @dongjoon-hyun :-)

**Pull request closed without merging**
